spring-kafka/src/main/java/org/springframework/kafka/listener/ListenerContainerPauseService.java

```diff
@@ -81,13 +82,37 @@ public void pause(MessageListenerContainer messageListenerContainer, Duration pa
 		}
 		else {
 			Instant resumeAt = Instant.now().plusMillis(pauseDuration.toMillis());
-			LOGGER.debug(() -> "Pausing container " + messageListenerContainer + "resume scheduled for "
+			LOGGER.debug(() -> "Pausing container " + messageListenerContainer + ", resume scheduled for "
 					+ resumeAt.atZone(ZoneId.systemDefault()).toLocalDateTime());
 			messageListenerContainer.pause();
-			this.scheduler.schedule(() -> resume(messageListenerContainer), resumeAt);
+			this.scheduler.schedule(() -> {
+				LOGGER.debug(() -> "Pausing container " + messageListenerContainer);
+				resume(messageListenerContainer);
+			}, resumeAt);
 		}
 	}
 
+	/**
+	 * Pause consumption from a given partition for the duration.
+	 * @param messageListenerContainer the container.
+	 * @param partition the partition.
+	 * @param pauseDuration the duration.
+	 */
+	public void pausePartition(MessageListenerContainer messageListenerContainer, TopicPartition partition,
+			Duration pauseDuration) {
+
+		Instant resumeAt = Instant.now().plusMillis(pauseDuration.toMillis());
+		LOGGER.debug(() -> "Pausing container: " + messageListenerContainer + " partition: " + partition
+				+ ", resume scheduled for "
+				+ resumeAt.atZone(ZoneId.systemDefault()).toLocalDateTime());
+		messageListenerContainer.pausePartition(partition);
+		this.scheduler.schedule(() -> {
+			LOGGER.debug(() -> "Resuming container: " + messageListenerContainer + " partition: " + partition);
+			messageListenerContainer.resumePartition(partition);
```

Comment:
In Spring Integration I had a bug in the past that the framework kept the memory consumption having references to heavy objects from scheduled tasks like this.
WDYT if we should avoid that over here as well in favor of simple listener container id?

Reply:
It's only a reference; the container is in memory anyway (in the container registry or application context). I think in SI, it was the persisted MG that was kept in memory unnecessarily.